app/src/lib/stores/repositories-store.ts

```diff
@@ -531,31 +557,100 @@ export class RepositoriesStore extends BaseStore {
     return record!.lastPruneDate
   }
 
-  public async isBranchProtected(
+  /**
+   * Inspect the cache or database to find out information about the branch
+   * protection configuration for the current repository and branch.
+   */
+  public async getBranchProtectionContext(
     gitHubRepository: GitHubRepository,
     branchName: string
-  ): Promise<boolean> {
+  ): Promise<BranchProtectionContext> {
     if (gitHubRepository.dbID === null) {
       return fatalError(
         'unable to get protected branches, GitHub repository has a null dbID'
       )
     }
 
-    const key = getKey(gitHubRepository.dbID, branchName)
+    const repoID = gitHubRepository.dbID
+
+    const branchProtectionsFound = this.branchProtectionSettingsFoundCache.get(
+      repoID
+    )
+
+    if (branchProtectionsFound === undefined) {
+      return this.loadAndCacheBranchProtection(repoID, branchName)
+    } else if (branchProtectionsFound) {
+      // as we know branch protections are enabled for the repository, check
+      // this specific branch either in the cache or the database
+      const isRemoteBranchProtected = await this.isBranchProtected(
+        repoID,
+        branchName
+      )
+
+      return {
+        branchProtectionsFound,
+        isRemoteBranchProtected,
+      }
+    }
+
+    // we have no branch protections enabled for this repository, therefore the
+    // current branch cannot be protected
+    return {
+      branchProtectionsFound,
+      isRemoteBranchProtected: false,
+    }
+  }
+
+  /**
+   * Load the branch protection information for a repository from the database
+   * and cache the results in memory
+   */
+  private async loadAndCacheBranchProtection(
+    repoID: number,
+    branchName: string
+  ) {
+    // query the database to find any protected branches
+    const branches = await this.db.protectedBranches
```

Comment:
i'm guessing this is not super fast based on our other performance investigations. the caching should help a lot with that such that this should only slow down the first commit in each repository per run of the app, correct?

Reply:
Correct. Ideally this should already be cached when `RepositoriesStore.updateGitHubRepository` was run, but for the edge case of a user making a commit immediately after starting (before that API refresh kicks in), it may not be cached - so we'll need to check the database.

And I realise now in this case we don't cache the `branchProtectionsFound` boolean in memory here, just the branch entries found. I'll address that.

Reply:
Addressed in 6f8c908a4

Reply:
I poked at this using `nodejs/node` (which has 12 protected branches) and the time to run this query was only **3ms** on my macOS machine, so I'm fine with the overhead here when we need to do this.

But given I already had caching in from the previous PR it made sense, I think sticking to this pattern is fine given this needs to run when committing - and also be low-overhead.

Going to mark this as resolved but let me know if you have any other questions about this.

Reply:
thanks for the extra info!